android_app/app/src/main/java/com/health/openscale/core/bluetooth/BluetoothActiveEraBF06.java

```diff
@@ -0,0 +1,326 @@
+            return;
+        } else if (pkt[0] != MAGIC_BYTE) {
+            Timber.w("Wrong packet MAGIC");
+            return;
+        } else if (pkt.length != 20) {
+            Timber.w("Wrong packet length %s expected 20", pkt.length);
+            return;
+        }
+
+        int packetType = pkt[0x12] & 0xFF;
+        switch (packetType) {
+            case 0xD5: // weight measurement
+                byte flags = pkt[0x02];
+                boolean stabilized = isBitSet(flags, 8);
+                isSupportHR = isBitSet(flags, 2);
+                isSupportPH = isBitSet(flags, 3);
+
+                float weightKg = (Converters.fromUnsignedInt24Be(pkt, 3) & 0x3FFFF) / 1000.0f;
+                // TODO: test if it's always in grams ?
+                if (stabilized && !weightStabilized) {
+                    weightStabilized = true;
+                    stableWeightKg = weightKg;
+                    Timber.i("Measured weight (stable): %.3f", stableWeightKg);
+                    scaleData.setWeight(weightKg);
+                    resumeMachineState();
+                }
+
+                break;
+
+            case 0xD0: // balance measuring
+                byte state = pkt[0x02];
+                boolean isFinal = state == 0x01;
+
+                int weightLRaw = Converters.fromUnsignedInt16Be(pkt, 3);
+                int percentLRaw = Converters.fromUnsignedInt16Be(pkt, 5);
+                float weightL = (float)weightLRaw / 100.0f;
+                float percentL = (float)percentLRaw / 10.0f;
+
+                if (isFinal && !balanceStabilized) {
+                    balanceStabilized = true;
+                    stableBalanceL = percentL;
+                    Timber.i("Measured balance (stable): L %.1f R: %.1f [%.2f]", percentL, 100.0f - percentL, weightL);
+                    resumeMachineState();
+                }
+                break;
+
+            case 0xD6: // reporting ADCs
+                byte number = pkt[0x02];
+                if (number == 1) {
+                    double imp = Converters.fromUnsignedInt16Be(pkt, 4);
+                    if (imp >= 1500.0d) {
+                        imp = (((imp  - 1000.0d) + ((stableWeightKg * 10.0d) * (-0.4d))) / 0.6d) / 10.0d;
+                    }
+                    impedance = imp;
+                    Timber.i("Measured impedance: %.1f", impedance);
+
+                    // calculate BIA using measure weight and impedance
+                    if (impedance > 0.0) {
+                        final ScaleUser selectedUser = OpenScale.getInstance().getSelectedScaleUser();
+                        int height = (int) Math.ceil(selectedUser.getBodyHeight());
+                        int age = selectedUser.getAge();
+                        int gender = selectedUser.getGender() == Converters.Gender.FEMALE ? 0 : 1;
+
+                        calculateBIA(height, impedance, stableWeightKg, age, gender);
+                        // TODO: report results
```

Comment:
you could set scaleData with the results?

Reply:
Here I've used example BIA calculation, just to see how far off the result whould be from what the scales are showing. The results are way off, so this is why I'm not storing the results

Reply:
ok yeah the vendors calculations are sometimes mysterious. I also implemented some estimates calculations, see https://github.com/oliexdev/openScale/wiki/Body-metric-estimations
You can enable them under settings->measurements->Body fat / Body water / Lean body mass